<a href="https://colab.research.google.com/github/tardebut/maskrcnnpy/blob/master/M1_Assignment_Stock_Price_Prediction_with_LLM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Module 1 No Code/Low Code Assignment: Stock Price Prediction with LLM (Using Gemini in Colab)

This assignment explores how large language models (LLMs) can perform complex tasks that previously required coding. Instead of manually writing code, we will interact with Google's Gemini (Click on the Gemini icon on the top right corner) in Google Colab to:

- Load and analyze stock price data.
- Engineer relevant features for prediction.
- Build and evaluate models such as regression and random forest.
- Discuss potential trading strategies based on predictions.

**Note**: While this notebook contains example code, students are encouraged to use LLMs to generate code snippets and explanations, managing interactions within the LLM's context window.

## Task Breakdown

### Step 1: Data Exploration
- **Prompt the LLM**: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance.
- **Key Considerations**: Ensure the model retains context while summarizing key statistics.

### Step 2: Feature Engineering
- **Prompt the LLM**: Generate meaningful features such as moving averages, technical indicators, and volatility measures.
- **Managing Context**: If the LLM forgets prior outputs, remind it by reintroducing key statistics.

### Step 3: Model Building
- **Prompt the LLM**: Train a model (regression for price prediction, random forest for trend classification).
- **Split Ratio Consideration**: Explain the importance of train-test split (e.g., 70/30) and verify the LLM respects it.

### Step 4: Model Evaluation & Prediction
- **Prompt the LLM**: Use the trained model to predict stock prices for 2023-2024.
- **Analyze Performance**: Interpret accuracy, R² scores, and classification reports.

### Step 5: Strategy Discussion
- **Prompt the LLM**: Based on model performance, discuss possible trading strategies.
- **Critical Thinking**: Evaluate whether the model is truly useful for financial decision-making.

---

Throughout the assignment, keep in mind that LLMs have a limited context window. If necessary, break down tasks into smaller chunks and reintroduce critical information.



## Using Gemini in Colab

To use Gemini directly in your browser within Google Colab:

1. **Create a new code cell** by clicking the `+ Code` button.
2. In the newly created cell, you will see a placeholder text: **"Start coding or generate with AI"**.
3. Click on **"Generate"** and enter your prompt, such as:
   > "Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics."
4. Gemini will generate code based on your prompt. You can then review, modify, and run it within the notebook.
5. If the model loses context, you may need to remind it of previous responses or break down tasks into smaller steps.

This method allows you to interactively generate and refine code without manually writing every line.


In [ ]:
# prompt: Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 using Yahoo Finance and summarize its key statistics.

!pip install yfinance

import yfinance as yf
import pandas as pd

# Download historical data for AAPL from 2020-01-01 to 2023-01-01
data = yf.download("AAPL", start="2020-01-01", end="2023-01-01")

# Summarize key statistics
print(data.describe())


YF.download() has changed argument auto_adjust default to True


[*********************100%***********************]  1 of 1 completed

Price        Close        High         Low        Open        Volume
Ticker        AAPL        AAPL        AAPL        AAPL          AAPL
count   756.000000  756.000000  756.000000  756.000000  7.560000e+02
mean    127.800474  129.350921  126.136626  127.722421  1.120920e+08
std      30.461355   30.739156   30.179073   30.489492  5.602586e+07
min      54.449894   55.452160   51.595990   55.350235  3.519590e+07
25%     112.857315  114.354702  111.397715  113.037711  7.636470e+07
50%     133.185188  134.319768  131.222483  133.006499  9.493580e+07
75%     148.415882  149.584513  146.468716  147.782312  1.296327e+08
max     178.879913  179.793905  176.039592  179.489238  4.265100e+08


## Now try doing step 2 - step 5 by yourself

The sample prompts and LLM responses are listed below for your reference.

# Sample Prompts and LLM Responses
---

## Step 1: Data Exploration

### Prompt:
> Load historical stock price data for Apple Inc. (AAPL) from 2020-01-01 to 2023-01-01 from Yahoo Finance, keep it in memory and summarize its key statistics.

### LLM Response:
- **Highest Closing Price:** \$182.01 on January 3, 2022
- **Lowest Closing Price:** \$56.01 on March 23, 2020
- **Average Closing Price:** \$126.69
- **Total Trading Days:** 756

---

## Step 2: Feature Engineering

### Prompt:
> Suggest some features that might help improve a stock price prediction model.

### LLM Response:
1. **Historical Prices**: Include past closing, opening, high, and low prices.
2. **Volume**: Daily trading volumes.
3. **Moving Averages**: Simple and exponential moving averages (e.g., 5-day, 10-day, 50-day).
4. **Technical Indicators**: RSI, MACD, Bollinger Bands, etc.
5. **Volatility**: Historical volatility, Average True Range (ATR).
6. **Sentiment Analysis**: News articles, social media sentiment.
7. **Economic Indicators**: Interest rates, inflation rates, GDP growth.
8. **Time-based Features**: Day of the week, month, year.
9. **Sector Performance**: Comparison with industry and sector indices.
10. **External Events**: Major geopolitical events, earnings reports.

---

## Step 3.1: Model Building - Linear Regression

### Prompt:
> Adopt the first four kinds of features and train a regression model to predict tomorrow's stock price. Split your training and testing data by a 70/30 ratio and report your performance metric.

### LLM Response:
- **Mean Squared Error (MSE):** 859.04
- **R-squared (R²) Score:** -0.009

These metrics indicate that the model is not performing well, as an R² score close to zero or negative suggests that the model does not explain the variability in the target variable better than a simple mean of the target variable. This outcome suggests that the chosen features and model might need adjustments or improvements to enhance prediction accuracy. ​​

---
(It might be too hard to predict the exact price but how about ups and downs?)
# Step 3.2: Model Building - Random Forest Classification

### Prompt:
> Create a binary target variable based on historical prices (1 if the close price increase tomorrow, 0 other), and train a Random Forest model using the same train/test split ratio. Report feature importance and performance metrics.

### LLM Response:
- **Accuracy:** 71.82%
- **Confusion Matrix:** [[84, 26],
 [36, 74]]
- **Classification Report:**

              precision    recall  f1-score   support

         0       0.70      0.76      0.73       110
         1       0.74      0.67      0.70       110
        accuracy                     0.72       220
        macro avg    0.72      0.72      0.72       220
        weighted avg  0.72      0.72      0.72       220



- **Feature Importance:** The most important feature for predicting whether the price will increase or decrease the next day is the closing price, followed by the 5-day moving average (MA_5). The model achieved an accuracy of approximately 71.82%, indicating a reasonable performance for this classification task.
         Feature	Importance
         Open	   0.0900
         High	   0.0883
         Low	   0.0980
         Close	   0.3087
         Volume	    0.0982
         MA_5	   0.1242
         MA_10	   0.0974
         MA_50	   0.0951

---

## Step 4: Model Evaluation & Prediction
### Prompt:
> Use this model to predict stock price increase/decrease for period starting from 2023-01-01 to 2024-01-01. Report your performance metrics.

### LLM Response:
The Random Forest classifier's performance on the new test data from 2023-01-01 to 2024-01-01 is as follows:
- **Accuracy:** 60.85%
- **Confusion Matrix:** [[105,   0],
 [ 83,  24]]
- **Classification Report:**

            precision    recall  f1-score   support

       0       0.56      1.00      0.72       105
       1       1.00      0.22      0.37       107

        accuracy                   0.61       212
        macro avg  0.78     0.61       0.54       212
        weighted avg 0.78    0.61        0.54       212

---
## Step 5: Strategy Discussion
### Prompt:
> What do you think about the performance of the model and what trading strategies could I build with this model?

### LLM Response:
- The model shows a moderate performance with an accuracy of approximately 60.85%. However, the precision and recall metrics indicate that while the model is good at identifying days when the stock price does not increase (class 0), it struggles significantly to identify days when the stock price does increase (class 1). This suggests a need for further feature engineering or model tuning to improve the prediction performance for this specific classification task.
- Potential Trading Strategies
Given the model's performance, potential trading strategies could be formulated as follows:

#### Conservative Strategy:

- **Focus**: Only trade on days when the model predicts a price increase with high confidence.
- **Action**: Buy stocks when the model predicts an increase (class 1).
- **Risk Management**: Since the model has a high precision for the no-increase class (class 0), it might be more reliable to stay out of the market when the model predicts no increase.

#### Contrarian Strategy:

- **Focus**: Take advantage of the model's high recall for no price increase predictions (class 0).
- **Action**: Sell short or avoid buying stocks on days when the model predicts no increase.
- **Risk Management**: This strategy assumes that the model’s predictions of no price increase are more reliable. It requires a careful assessment of market conditions to avoid false signals.

#### Hybrid Strategy:

- **Combine**: Use both conservative and contrarian strategies based on confidence thresholds.
- **Action**: Set a higher threshold for buying (only buy when the predicted probability of price increase is significantly high) and a lower threshold for selling or staying out of the market.
- **Risk Management**: Adjust thresholds dynamically based on recent model performance and market volatility.


---
## Conclusion

Through this assignment, we have explored how LLMs can assist in complex machine learning tasks without requiring extensive coding knowledge. Key takeaways:

- **Context Management**: LLMs have limited memory, so breaking tasks into smaller steps and reinforcing context is crucial.
- **Feature Engineering with LLMs**: The model can suggest features but needs guidance on financial relevance.
- **Modeling with LLMs**: LLMs can write and refine code, but human judgment is necessary to validate results.
- **Real-world Limitations**: Stock price prediction remains challenging; models should be used critically and not solely relied upon for trading decisions.

Next Steps:
- Experiment with different LLMs and compare outputs.
- Try refining feature engineering techniques.
- Test alternative modeling approaches, such as deep learning or ensemble methods.



# Task
Download the Excel file from the provided Google Sheet URL and load its content into a pandas DataFrame, then print the head of the DataFrame to confirm successful loading.

## Download and Load Google Sheet Data

### Subtask:
Generate Python code to download the Excel file from the provided Google Sheet URL and load its content into a pandas DataFrame. This will involve converting the shared link into a direct download URL and then using pandas to read the Excel data. It will also print the head of the DataFrame to confirm successful loading.


**Reasoning**:
To begin, I will provide a Python code block that includes a placeholder for the Google Sheet URL. This code will demonstrate how to convert a Google Sheet sharing link into a direct download URL for an Excel file and then load that file into a pandas DataFrame. Finally, it will display the first few rows of the DataFrame to confirm successful data loading.



In [3]:
import pandas as pd

# Replace with your actual Google Sheet URL
# Ensure the Google Sheet is shared with 'Anyone with the link' set to 'Viewer' or higher.
google_sheet_url = 'https://docs.google.com/spreadsheets/d/1d92qF6iD6-XlO6B2bP0NlHjTVfWpUiUM/edit?usp=sharing'

# Convert the sharing URL to a direct download URL for Excel
# This typically involves replacing 'edit?usp=sharing' with 'export?format=xlsx'
excel_download_url = google_sheet_url.replace('/edit?usp=sharing', '/export?format=xlsx')

# Read the Excel file directly into a pandas DataFrame
try:
    df = pd.read_excel(excel_download_url)
    print("Successfully loaded data into DataFrame.")
    print("First 5 rows of the DataFrame:")
    print(df.head())
except Exception as e:
    print(f"An error occurred: {e}")
    print("Please ensure the Google Sheet URL is correct and shared with 'Anyone with the link'.")

Successfully loaded data into DataFrame.
First 5 rows of the DataFrame:
   District  tpk6  t712  texep  telse  ttotal  Guidance  socialw  pschol  \
0   ALACHUA   776   666    267     56    1765        48       14       9   
1     BAKER   145   113     28      7     293        13        0       2   
2       BAY   660   648    301    194    1803        65        6       8   
3  BRADFORD    82    62     37      7     188         6        0       1   
4   BREVARD  2096  1622    816     71    4605       171       46      26   

   librarians  otherinstruct  yearsexp    salary  
0          43            114     11.43  47853.72  
1           5             16     11.81  43567.38  
2          31            146     11.65  45615.83  
3           1             19      8.31  41776.51  
4          84            442     13.67  46550.16  


In [4]:
if 'df' in locals():
    print("First 5 rows of the DataFrame:")
    display(df.head())
else:
    print("DataFrame 'df' is not yet loaded. Please ensure cell f1f036ec was executed successfully.")

First 5 rows of the DataFrame:


,District,tpk6,t712,texep,telse,ttotal,Guidance,socialw,pschol,librarians,otherinstruct,yearsexp,salary
0,ALACHUA,776,666,267,56,1765,48,14,9,43,114,11.43,47853.72
1,BAKER,145,113,28,7,293,13,0,2,5,16,11.81,43567.38
2,BAY,660,648,301,194,1803,65,6,8,31,146,11.65,45615.83
3,BRADFORD,82,62,37,7,188,6,0,1,1,19,8.31,41776.51
4,BREVARD,2096,1622,816,71,4605,171,46,26,84,442,13.67,46550.16


It appears the previous attempt to load the data failed because the `google_sheet_url` was a placeholder (`YOUR_SHEET_ID`) and resulted in a '404 Not Found' error.

**Please update the `google_sheet_url` variable in the code block above with your actual Google Sheet URL.**

Make sure your Google Sheet is shared with 'Anyone with the link' set to 'Viewer' or higher, and then re-run the code block. Once you have replaced the placeholder and executed the cell, the data should load successfully into a pandas DataFrame.

It appears the previous attempt to load the data failed because the `google_sheet_url` was a placeholder (`YOUR_SHEET_ID`) and resulted in a '404 Not Found' error.

**Please update the `google_sheet_url` variable in the code block above with your actual Google Sheet URL.**

Make sure your Google Sheet is shared with 'Anyone with the link' set to 'Viewer' or higher, and then re-run the code block. Once you have replaced the placeholder and executed the cell, the data should load successfully into a pandas DataFrame.

## Download and Load Google Sheet Data (Retry)

### Subtask:
Correct the Google Sheet URL and load the Excel file content into a pandas DataFrame, then print the head of the DataFrame to confirm successful loading.


### Instructions to Correct Google Sheet URL:

The previous code cell (cell_id `f1f036ec`) attempted to download an Excel file from a Google Sheet URL, but it failed because the URL provided was a placeholder.

To proceed, please follow these steps:

1.  **Go to the code cell with ID `f1f036ec`**. This is the cell containing the Python code for loading the Google Sheet.
2.  **Locate the line**: `google_sheet_url = 'https://docs.google.com/spreadsheets/d/YOUR_SHEET_ID/edit?usp=sharing'`
3.  **Replace the placeholder URL** with the actual sharing link of your Google Sheet.
    *   **Crucial**: Ensure your Google Sheet's sharing settings are set to 'Anyone with the link' with at least 'Viewer' access. Without this, the download will fail even with a correct URL.
4.  **After updating the URL, re-run the code cell `f1f036ec`**.

Upon successful execution, you should see "Successfully loaded data into DataFrame." and the first 5 rows of your DataFrame printed as output.

## Final Task

### Subtask:
Acknowledge the status of the Google Sheet data loading and provide next steps for the user.


## Summary:

### Data Analysis Key Findings
*   The initial attempt to load data from a Google Sheet failed due to an `HTTP Error 404: Not Found`.
*   This error occurred because the provided `google_sheet_url` was a placeholder (`'https://docs.google.com/spreadsheets/d/YOUR_SHEET_ID/edit?usp=sharing'`) and not a valid, actual Google Sheet URL.
*   The system provided the correct Python code structure to download and load an Excel file from a Google Sheet, given a valid URL.
*   The solution also included comprehensive instructions for the user on how to update the placeholder URL and ensure proper sharing settings for the Google Sheet.

### Insights or Next Steps
*   The user must replace the placeholder `google_sheet_url` with the actual sharing link of their Google Sheet.
*   It is crucial to ensure that the Google Sheet's sharing settings are configured to 'Anyone with the link' with at least 'Viewer' access for the download to succeed.
